File: 백준/Gold/17425. 약수의 합/약수의 합.py
import sys
input = sys.stdin.readline
sum_arr = [0 for _ in range(1000001)] #누적합
dp = [1 for _ in range(1000001)] #dp 1

for i in range(2, 1000001):
    j = 1
    while i * j <= 1000000:
        # i의 배수는 약수로 i를 가지고 있으므로 해당 i의 배수 인덱스에 약수 i의 값을 더해줌.
        dp[i*j] += i 
        j += 1
for i in range(1, 1000001):
    sum_arr[i] = sum_arr[i-1] + dp[i]

t = int(input())
for _ in range(t):
    n = int(input())    
    print(sum_arr[n])

File: 백준/Gold/17425. 약수의 합/약수의 합.java

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		// 약수의 배수들의 합
		long[] dp = new long[1000001];
		Arrays.fill(dp, 1);
		
		// 누적합
		long[] sum_arr = new long[1000001];
		
		// 약수의 배수들의 합
		for(int i = 2; i<=1000000; i++) {
			int j = 1;
			while (i*j <= 1000000) {
				dp[i*j] += i;
				j += 1;
			}
		}
		
		// 누적합
		for (int i = 1; i <= 1000000; i++) {
			sum_arr[i] = sum_arr[i-1] + dp[i];
		}
		
		int T = Integer.parseInt(br.readLine());
		// 테스트 케이스
		for(int i = 0; i < T; i++) {
			int N = Integer.parseInt(br.readLine());
			sb.append(sum_arr[N]).append("\n");
		}
		System.out.println(sb);
	}
}
